File: tests/mpesa_parser.js
/**
 * Kenyan M-PESA SMS Deterministic Parser
 * High precision extractor for M-PESA transactions
 */

export class MpesaParser {
  /**
   * Parse an incoming SMS message text
   * @param {string} smsBody
   * @returns {object|null} Parsed transaction or null if not an M-PESA message
   */
  static parse(smsBody) {
    if (!smsBody || typeof smsBody !== 'string') {
      return null;
    }

    const trimmed = smsBody.trim();

    // Must contain Confirmed and an 8-12 char alphanumeric transaction code at start
    const codeMatch = trimmed.match(/^([A-Z0-9]{8,12})\s+Confirmed\./i);
    if (!codeMatch) {
      return {
        isFinancial: false,
        confidence: 0,
        rawText: trimmed,
        error: 'Not a standard confirmed M-PESA financial transaction'
      };
    }

    const transactionCode = codeMatch[1].toUpperCase();

    // 1. Extract Balance if present
    let balance = null;
    const balanceMatch = trimmed.match(/New M-PESA balance is Ksh([\d,]+\.?\d*)/i);
    if (balanceMatch) {
      balance = parseFloat(balanceMatch[1].replace(/,/g, ''));
    }

    // 2. Extract Transaction Cost if present
    let cost = null;
    const costMatch = trimmed.match(/Transaction cost,?\s*Ksh([\d,]+\.?\d*)/i);
    if (costMatch) {
      cost = parseFloat(costMatch[1].replace(/,/g, ''));
    }

    // 3. Pattern: Sent to Paybill / Buy Goods / Person
    // e.g. "Ksh3,500.00 sent to Kenya Power and Lighting Company for account 12345678 on 7/9/26 at 8:15 PM."
    // e.g. "Ksh2,000.00 sent to JANE DOE 0712345678 on 7/9/26 at 10:05 AM."
    const sentMatch = trimmed.match(/Ksh([\d,]+\.?\d*)\s+sent to\s+(.+?)\s+on\s+(\d{1,2}\/\d{1,2}\/\d{2,4})\s+at\s+(\d{1,2}:\d{2}\s*(?:AM|PM))/i);
    if (sentMatch) {
      const amount = parseFloat(sentMatch[1].replace(/,/g, ''));
      let targetRaw = sentMatch[2].trim();
      const dateStr = sentMatch[3];
      const timeStr = sentMatch[4];

      let accountRef = null;
      const acctMatch = targetRaw.match(/^(.+?)\s+for account\s+(.+)$/i);
      if (acctMatch) {
        targetRaw = acctMatch[1].trim();
        accountRef = acctMatch[2].trim();
      }

      let phone = null;
      const phoneMatch = targetRaw.match(/^(.+?)\s+(07\d{8}|01\d{8}|\+?254\d{9})$/);
      if (phoneMatch) {
        targetRaw = phoneMatch[1].trim();
        phone = phoneMatch[2].trim();
      }

      const classification = classifySentTransaction(targetRaw, accountRef);

      return {
        isFinancial: true,
        confidence: 0.95,
        transactionCode,
        amount,
        balance,
        cost,
        date: formatDateIso(dateStr),
        rawDate: dateStr,
        time: timeStr,
        recipient: targetRaw,
        description: targetRaw + (accountRef ? ` (${accountRef})` : ''),
        accountRef,
        phone,
        type: classification.type,
        category: classification.category,
        account: 'Mpesa',
        destinationAccount: classification.destinationAccount || null,
        rawText: trimmed
      };
    }

    // 4. Pattern: Paid to Till / Buy Goods
    // e.g. "Ksh1,250.00 paid to NAIVAS SUPERMARKET. on 7/9/26 at 2:30 PM."
    const paidMatch = trimmed.match(/Ksh([\d,]+\.?\d*)\s+paid to\s+(.+?)(?:\.|\s+on)\s+(?:on\s+)?(\d{1,2}\/\d{1,2}\/\d{2,4})\s+at\s+(\d{1,2}:\d{2}\s*(?:AM|PM))/i);
    if (paidMatch) {
      const amount = parseFloat(paidMatch[1].replace(/,/g, ''));
      let merchant = paidMatch[2].trim().replace(/\.$/, '');
      const dateStr = paidMatch[3];
      const timeStr = paidMatch[4];

      const classification = classifyPaidMerchant(merchant);

      return {
        isFinancial: true,
        confidence: 0.95,
        transactionCode,
        amount,
        balance,
        cost,
        date: formatDateIso(dateStr),
        rawDate: dateStr,
        time: timeStr,
        recipient: merchant,
        description: merchant,
        type: classification.type,
        category: classification.category,
        account: 'Mpesa',
        destinationAccount: null,
        rawText: trimmed
      };
    }

    // 5. Pattern: Received Money
    // e.g. "You have received Ksh50,000.00 from ACME TECH LTD 0798765432 on 7/9/26 at 9:00 AM."
    const receivedMatch = trimmed.match(/You have received Ksh([\d,]+\.?\d*)\s+from\s+(.+?)\s+on\s+(\d{1,2}\/\d{1,2}\/\d{2,4})\s+at\s+(\d{1,2}:\d{2}\s*(?:AM|PM))/i);
    if (receivedMatch) {
      const amount = parseFloat(receivedMatch[1].replace(/,/g, ''));
      let senderRaw = receivedMatch[2].trim();
      const dateStr = receivedMatch[3];
      const timeStr = receivedMatch[4];

      let phone = null;
      const phoneMatch = senderRaw.match(/^(.+?)\s+(07\d{8}|01\d{8}|\+?254\d{9})$/);
      if (phoneMatch) {
        senderRaw = phoneMatch[1].trim();
        phone = phoneMatch[2].trim();
      }

      return {
        isFinancial: true,
        confidence: 0.95,
        transactionCode,
        amount,
        balance,
        cost: 0,
        date: formatDateIso(dateStr),
        rawDate: dateStr,
        time: timeStr,
        sender: senderRaw,
        description: `Received from ${senderRaw}`,
        phone,
        type: 'Income',
        category: 'Salary', // Default suggestion, reviewed by user
        account: 'Mpesa',
        destinationAccount: null,
        rawText: trimmed
      };
    }

    // 6. Pattern: Cash Withdrawal from Agent
    // e.g. "on 7/9/26 at 4:45 PM Withdraw Ksh3,000.00 from 12345 - AGENT STORE"
    const withdrawMatch = trimmed.match(/on\s+(\d{1,2}\/\d{1,2}\/\d{2,4})\s+at\s+(\d{1,2}:\d{2}\s*(?:AM|PM))\s+Withdraw\s+Ksh([\d,]+\.?\d*)\s+from\s+(.+?)(?:\s+New|\.|$)/i);
    if (withdrawMatch) {
      const dateStr = withdrawMatch[1];
      const timeStr = withdrawMatch[2];
      const amount = parseFloat(withdrawMatch[3].replace(/,/g, ''));
      const agent = withdrawMatch[4].trim();

      return {
        isFinancial: true,
        confidence: 0.90,
        transactionCode,
        amount,
        balance,
        cost,
        date: formatDateIso(dateStr),
        rawDate: dateStr,
        time: timeStr,
        recipient: agent,
        description: `Withdrawal: ${agent}`,
        type: 'Transfer',
        category: 'Internal Account Transfer',
        account: 'Mpesa',
        destinationAccount: 'Cash',
        rawText: trimmed
      };
    }

    // Unmatched pattern despite Confirmed code -> Flag for manual review
    return {
      isFinancial: true,
      confidence: 0.40,
      transactionCode,
      amount: null,
      balance,
      date: new Date().toISOString().split('T')[0],
      type: 'Expenses',
      category: 'Uncategorized',
      account: 'Mpesa',
      destinationAccount: null,
      rawText: trimmed,
      note: 'Ambiguous SMS pattern: manual input required'
    };
  }
}

/**
 * Classifies paybill / send money targets into categories & types
 */
function classifySentTransaction(recipient, accountRef) {
  const r = (recipient || '').toLowerCase();
  const a = (accountRef || '').toLowerCase();

  // 1. Debt & Loans (check loan keywords first to avoid Sacco vs Sacco Loan collision)
  if (r.includes('equity loan') || (r.includes('equity') && (r.includes('loan') || a.includes('loan')))) {
    return { type: 'Debt', category: 'Equity Loan' };
  }
  if (r.includes('nca sacco loan') || (r.includes('nca') && (r.includes('loan') || a.includes('loan')))) {
    return { type: 'Debt', category: 'NCA Sacco Loan' };
  }
  if (r.includes('helb')) {
    return { type: 'Debt', category: 'Helb Loan' };
  }
  if (r.includes('fuliza') || r.includes('m-shwari') || r.includes('tala') || r.includes('branch')) {
    return { type: 'Debt', category: 'Equity Loan' };
  }

  // 2. Specific Savings Institutions
  // Tower Sacco: Payment to Tower Sacco via SMS is a Savings contribution funded from Mpesa
  if (r.includes('tower sacco') || r.includes('tower')) {
    return { type: 'Savings', category: 'Tower Sacco', destinationAccount: null };
  }
  // NCA Sacco: Payment to NCA Sacco (without loan) is a Savings contribution
  if (r.includes('nca sacco') || r.includes('nca')) {
    return { type: 'Savings', category: 'NCA Sacco', destinationAccount: null };
  }
  if (r.includes('sanlam')) {
    return { type: 'Savings', category: 'Sanlam MMF', destinationAccount: null };
  }
  if (r.includes('britam')) {
    return { type: 'Savings', category: 'Britam EQ and MMF', destinationAccount: null };
  }
  if (r.includes('etica')) {
    return { type: 'Savings', category: 'Etica MMF', destinationAccount: null };
  }
  if (r.includes('ziidi')) {
    return { type: 'Savings', category: 'Ziidi MMF', destinationAccount: null };
  }
  if (r.includes('cic') || r.includes('mmf') || r.includes('money market')) {
    return { type: 'Savings', category: 'Money Market Fund (MMF)', destinationAccount: 'MMF Account' };
  }

  // 3. Matatu Transport Saccos (Expenses -> Fare)
  if (r.includes('super metro') || r.includes('2nk') || r.includes('lopha') || r.includes('kbs') || r.includes('city hoppa') || r.includes('metro') || r.includes('matatu')) {
    return { type: 'Expenses', category: 'Fare' };
  }

  // 4. Utilities / Bills
  if (r.includes('kenya power') || r.includes('kplc')) {
    return { type: 'Bills', category: 'Electricity / KPLC' };
  }
  if (r.includes('nairobi water') || r.includes('water')) {
    return { type: 'Bills', category: 'Water' };
  }
  if (r.includes('safaricom home') || r.includes('zuku') || r.includes('faiba') || r.includes('poa')) {
    return { type: 'Bills', category: 'Internet / WiFi' };
  }

  // 5. Bank Transfers
  if (r.includes('ncba loop') || r.includes('loop')) {
    return { type: 'Transfer', category: 'Internal Account Transfer', destinationAccount: 'Bank (NCBA Loop)' };
  }
  if (r.includes('equity') || r.includes('equity bank')) {
    return { type: 'Transfer', category: 'Internal Account Transfer', destinationAccount: 'Bank (Equity)' };
  }
  if (r.includes('kcb') || r.includes('kcb bank')) {
    return { type: 'Transfer', category: 'Internal Account Transfer', destinationAccount: 'Bank (KCB)' };
  }
  if (r.includes('sacco') || r.includes('stima')) {
    return { type: 'Transfer', category: 'Internal Account Transfer', destinationAccount: 'SACCO Account' };
  }

  // Default P2P
  return { type: 'Expenses', category: 'Gifts / Support' };
}

/**
 * Classifies Buy Goods merchant names
 */
function classifyPaidMerchant(merchant) {
  const m = (merchant || '').toLowerCase();

  // Groceries / Supermarket
  if (m.includes('naivas') || m.includes('carrefour') || m.includes('quickmart') || m.includes('chandarana') || m.includes('cleanshelf')) {
    return { type: 'Expenses', category: 'Groceries' };
  }

  // Dining
  if (m.includes('kfc') || m.includes('java') || m.includes('artcaffe') || m.includes('pizza') || m.includes('restaurant') || m.includes('cafe')) {
    return { type: 'Expenses', category: 'Dining Out / Takeout' };
  }

  // Fuel / Transport
  if (m.includes('total') || m.includes('shell') || m.includes('rubis') || m.includes('ola') || m.includes('uber') || m.includes('bolt')) {
    return { type: 'Expenses', category: 'Transport & Fuel' };
  }

  // Health
  if (m.includes('pharmacy') || m.includes('chemist') || m.includes('hospital') || m.includes('clinic')) {
    return { type: 'Expenses', category: 'Health & Pharmacy' };
  }

  return { type: 'Expenses', category: 'Shopping & Clothing' };
}

/**
 * Formats "D/M/YY" or "DD/MM/YYYY" to "YYYY-MM-DD"
 */
function formatDateIso(dateStr) {
  if (!dateStr) return new Date().toISOString().split('T')[0];
  const parts = dateStr.split('/');
  if (parts.length !== 3) return dateStr;

  let day = parts[0].padStart(2, '0');
  let month = parts[1].padStart(2, '0');
  let year = parts[2];
  if (year.length === 2) {
    year = '20' + year;
  }
  return `${year}-${month}-${day}`;
}

File: android/app/src/main/java/com/gmdparser/parser/MpesaParser.kt
package com.gmdparser.parser

import com.gmdparser.data.model.Transaction
import com.gmdparser.data.model.TransactionStatus
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

object MpesaParser {

    private val CONFIRMED_CODE_REGEX = Regex("^([A-Z0-9]{8,12})\\s+Confirmed[\\.\\s]", RegexOption.IGNORE_CASE)
    private val BALANCE_REGEX = Regex("New M-PESA balance is (?:Ksh|KES)\\.?\\s*([\\d,]+\\.?\\d*)", RegexOption.IGNORE_CASE)
    private val COST_REGEX = Regex("Transaction cost,?\\s*(?:Ksh|KES)\\.?\\s*([\\d,]+\\.?\\d*)", RegexOption.IGNORE_CASE)

    // 1. Sent to (Paybill, Send Money, Bank)
    private val SENT_REGEX = Regex(
        "(?:Ksh|KES)\\.?\\s*([\\d,]+\\.?\\d*)\\s+sent to\\s+(.+?)\\s+on\\s+(\\d{1,2}[/-]\\d{1,2}[/-]\\d{2,4})\\s+at\\s+(\\d{1,2}:\\d{2}\\s*(?:AM|PM)?)",
        RegexOption.IGNORE_CASE
    )

    // 2. Paid to (Buy Goods / Till)
    private val PAID_REGEX = Regex(
        "(?:Ksh|KES)\\.?\\s*([\\d,]+\\.?\\d*)\\s+paid to\\s+(.+?)(?:\\.|\\s+on)\\s+(?:on\\s+)?(\\d{1,2}[/-]\\d{1,2}[/-]\\d{2,4})\\s+at\\s+(\\d{1,2}:\\d{2}\\s*(?:AM|PM)?)",
        RegexOption.IGNORE_CASE
    )

    // 3. Received from
    private val RECEIVED_REGEX = Regex(
        "You have received\\s+(?:Ksh|KES)\\.?\\s*([\\d,]+\\.?\\d*)\\s+from\\s+(.+?)\\s+on\\s+(\\d{1,2}[/-]\\d{1,2}[/-]\\d{2,4})\\s+at\\s+(\\d{1,2}:\\d{2}\\s*(?:AM|PM)?)",
        RegexOption.IGNORE_CASE
    )

    // 4. Withdrawal from agent
    private val WITHDRAW_REGEX = Regex(
        "on\\s+(\\d{1,2}[/-]\\d{1,2}[/-]\\d{2,4})\\s+at\\s+(\\d{1,2}:\\d{2}\\s*(?:AM|PM)?)\\s+Withdraw\\s+(?:Ksh|KES)\\.?\\s*([\\d,]+\\.?\\d*)\\s+from\\s+(.+?)(?:\\s+New|\\.|$)",
        RegexOption.IGNORE_CASE
    )

    // 5. Airtime / Bundles purchase
    private val AIRTIME_REGEX = Regex(
        "(?:You\\s+(?:have\\s+)?bought|bought)\\s+(?:Ksh|KES)\\.?\\s*([\\d,]+\\.?\\d*)\\s+of\\s+(?:airtime|data(?:\\s+bundles?)?|bundles)(?:\\s+for\\s+(.+?))?\\s+on\\s+(\\d{1,2}[/-]\\d{1,2}[/-]\\d{2,4})\\s+at\\s+(\\d{1,2}:\\d{2}\\s*(?:AM|PM)?)",
        RegexOption.IGNORE_CASE
    )

    fun parse(smsBody: String?): Transaction? {
        if (smsBody.isNullOrBlank()) return null
        val trimmed = smsBody.trim()

        val codeMatch = CONFIRMED_CODE_REGEX.find(trimmed) ?: return null
        val txCode = codeMatch.groupValues[1].uppercase(Locale.ROOT)

        val balance = BALANCE_REGEX.find(trimmed)?.groupValues?.get(1)?.replace(",", "")?.toDoubleOrNull()
        val cost = COST_REGEX.find(trimmed)?.groupValues?.get(1)?.replace(",", "")?.toDoubleOrNull()

        // 1. Sent to (Paybill, Send Money, Bank Transfer)
        val sentMatch = SENT_REGEX.find(trimmed)
        if (sentMatch != null) {
            val amount = sentMatch.groupValues[1].replace(",", "").toDoubleOrNull() ?: 0.0
            var targetRaw = sentMatch.groupValues[2].trim()
            val dateStr = sentMatch.groupValues[3]
            val timeStr = sentMatch.groupValues[4]

            var accountRef: String? = null
            val acctMatch = Regex("^(.+?)\\s+for account\\s+(.+)$", RegexOption.IGNORE_CASE).find(targetRaw)
            if (acctMatch != null) {
                targetRaw = acctMatch.groupValues[1].trim()
                accountRef = acctMatch.groupValues[2].trim()
            }

            var phone: String? = null
            val phoneMatch = Regex("^(.+?)\\s+(07\\d{8}|01\\d{8}|\\+?254\\d{9})$").find(targetRaw)
            if (phoneMatch != null) {
                targetRaw = phoneMatch.groupValues[1].trim()
                phone = phoneMatch.groupValues[2].trim()
            }

            val classification = classifySentTarget(targetRaw, accountRef)

            return Transaction(
                transactionCode = txCode,
                amount = amount,
                type = classification.type,
                category = classification.category,
                description = targetRaw + if (!accountRef.isNullOrBlank()) " ($accountRef)" else "",
                account = "Mpesa",
                date = formatIsoDate(dateStr),
                time = timeStr,
                destinationAccount = classification.destinationAccount,
                balance = balance,
                cost = cost,
                senderOrRecipient = targetRaw,
                rawText = trimmed,
                status = TransactionStatus.PENDING_REVIEW
            )
        }

        // 2. Paid to (Buy Goods / Till)
        val paidMatch = PAID_REGEX.find(trimmed)
        if (paidMatch != null) {
            val amount = paidMatch.groupValues[1].replace(",", "").toDoubleOrNull() ?: 0.0
            val merchant = paidMatch.groupValues[2].trim().removeSuffix(".")
            val dateStr = paidMatch.groupValues[3]
            val timeStr = paidMatch.groupValues[4]

            val classification = classifyMerchant(merchant)

            return Transaction(
                transactionCode = txCode,
                amount = amount,
                type = classification.type,
                category = classification.category,
                description = merchant,
                account = "Mpesa",
                date = formatIsoDate(dateStr),
                time = timeStr,
                destinationAccount = null,
                balance = balance,
                cost = cost,
                senderOrRecipient = merchant,
                rawText = trimmed,
                status = TransactionStatus.PENDING_REVIEW
            )
        }

        // 3. Received from (Income)
        val receivedMatch = RECEIVED_REGEX.find(trimmed)
        if (receivedMatch != null) {
            val amount = receivedMatch.groupValues[1].replace(",", "").toDoubleOrNull() ?: 0.0
            var sender = receivedMatch.groupValues[2].trim()
            val dateStr = receivedMatch.groupValues[3]
            val timeStr = receivedMatch.groupValues[4]

            val phoneMatch = Regex("^(.+?)\\s+(07\\d{8}|01\\d{8}|\\+?254\\d{9})$").find(sender)
            if (phoneMatch != null) {
                sender = phoneMatch.groupValues[1].trim()
            }

            return Transaction(
                transactionCode = txCode,
                amount = amount,
                type = "Income",
                category = "Salary",
                description = "Received from $sender",
                account = "Mpesa",
                date = formatIsoDate(dateStr),
                time = timeStr,
                destinationAccount = null,
                balance = balance,
                cost = 0.0,
                senderOrRecipient = sender,
                rawText = trimmed,
                status = TransactionStatus.PENDING_REVIEW
            )
        }

        // 4. Withdrawal from agent
        val withdrawMatch = WITHDRAW_REGEX.find(trimmed)
        if (withdrawMatch != null) {
            val dateStr = withdrawMatch.groupValues[1]
            val timeStr = withdrawMatch.groupValues[2]
            val amount = withdrawMatch.groupValues[3].replace(",", "").toDoubleOrNull() ?: 0.0
            val agent = withdrawMatch.groupValues[4].trim()

            return Transaction(
                transactionCode = txCode,
                amount = amount,
                type = "Balance",
                category = "",
                description = "Withdrawal: $agent",
                account = "Mpesa",
                date = formatIsoDate(dateStr),
                time = timeStr,
                destinationAccount = "Cash",
                balance = balance,
                cost = cost,
                senderOrRecipient = agent,
                rawText = trimmed,
                status = TransactionStatus.PENDING_REVIEW
            )
        }

        // 5. Airtime or Bundles purchase
        val airtimeMatch = AIRTIME_REGEX.find(trimmed)
        if (airtimeMatch != null) {
            val amount = airtimeMatch.groupValues[1].replace(",", "").toDoubleOrNull() ?: 0.0
            val recipient = airtimeMatch.groupValues[2].trim().ifEmpty { "Airtime" }
            val dateStr = airtimeMatch.groupValues[3]
            val timeStr = airtimeMatch.groupValues[4].trim()

            return Transaction(
                transactionCode = txCode,
                amount = amount,
                type = "Expenses",
                category = "Bundles",
                description = "Airtime for $recipient".trim(),
                account = "Mpesa",
                date = formatIsoDate(dateStr),
                time = timeStr,
                destinationAccount = null,
                balance = balance,
                cost = cost ?: 0.0,
                senderOrRecipient = recipient,
                rawText = trimmed,
                status = TransactionStatus.PENDING_REVIEW
            )
        }

        // 6. Smart fallback: extract amount, date, time and recipient from confirmed M-PESA SMS
        val fallbackAmountMatch = Regex("(?:Ksh|KES)\\.?\\s*([\\d,]+\\.?\\d*)", RegexOption.IGNORE_CASE).find(trimmed)
        val extractedAmount = fallbackAmountMatch?.groupValues?.get(1)?.replace(",", "")?.toDoubleOrNull() ?: 0.0

        val fallbackDateMatch = Regex("(\\d{1,2}[/-]\\d{1,2}[/-]\\d{2,4})").find(trimmed)
        val extractedDate = fallbackDateMatch?.groupValues?.get(1)?.let { formatIsoDate(it) }
            ?: SimpleDateFormat("yyyy-MM-dd", Locale.ROOT).format(Date())

        val fallbackTimeMatch = Regex("(\\d{1,2}:\\d{2}\\s*(?:AM|PM)?)", RegexOption.IGNORE_CASE).find(trimmed)
        val extractedTime = fallbackTimeMatch?.groupValues?.get(1)?.trim() ?: ""

        val payeeMatch = Regex("(?:sent to|paid to|from|bought)\\s+([A-Za-z0-9\\s\\.\\-]+?)(?:\\s+on|\\s+for account|\\s+New M-PESA|\\.|,|$)", RegexOption.IGNORE_CASE).find(trimmed)
        val extractedDesc = payeeMatch?.groupValues?.get(1)?.trim()?.ifEmpty { "M-PESA Transaction" } ?: "M-PESA Transaction"

        val classification = classifySentTarget(extractedDesc, null)

        return Transaction(
            transactionCode = txCode,
            amount = extractedAmount,
            type = classification.type,
            category = classification.category,
            description = extractedDesc,
            account = "Mpesa",
            date = extractedDate,
            time = extractedTime,
            destinationAccount = null,
            balance = balance,
            cost = cost,
            senderOrRecipient = extractedDesc,
            rawText = trimmed,
            status = TransactionStatus.PENDING_REVIEW
        )
    }

    fun createManualBalanceTransaction(
        sourceAccount: String,
        destinationAccount: String,
        amount: Double,
        transactionCode: String? = null,
        date: String? = null
    ): Transaction {
        val code = transactionCode ?: "BAL-${System.currentTimeMillis()}"
        val today = date ?: SimpleDateFormat("yyyy-MM-dd", Locale.ROOT).format(Date())
        return Transaction(
            transactionCode = code,
            amount = amount,
            type = "Balance",
            category = "",
            description = "Transfer from $sourceAccount to $destinationAccount",
            account = sourceAccount,
            destinationAccount = destinationAccount,
            date = today,
            status = TransactionStatus.PENDING_REVIEW
        )
    }

    private data class Classification(val type: String, val category: String, val destinationAccount: String? = null)

    private fun classifySentTarget(target: String, accountRef: String?): Classification {
        val t = target.lowercase(Locale.ROOT)
        val a = accountRef?.lowercase(Locale.ROOT) ?: ""

        // 1. Debt & Loans
        if (t.contains("equity loan") || (t.contains("equity") && (t.contains("loan") || a.contains("loan")))) {
            return Classification("Debt", "Equity Loan")
        }
        if (t.contains("nca sacco loan") || (t.contains("nca sacco") && (t.contains("loan") || a.contains("loan")))) {
            return Classification("Debt", "NCA Sacco Loan")
        }
        if (t.contains("helb")) {
            return Classification("Debt", "Helb Loan")
        }
        if (t.contains("fuliza") || t.contains("m-shwari") || t.contains("tala") || t.contains("branch")) {
            return Classification("Debt", "Equity Loan")
        }

        // 2. Specific Savings Institutions
        // Tower Sacco: Payment to Tower Sacco via SMS is a Savings contribution funded from Mpesa
        if (t.contains("tower sacco")) {
            return Classification("Savings", "Tower Sacco", null)
        }
        // NCA Sacco: Payment to NCA Sacco (without loan) is a Savings contribution
        if (t.contains("nca sacco")) {
            return Classification("Savings", "NCA Sacco", null)
        }
        if (t.contains("sanlam")) {
            return Classification("Savings", "Sanlam MMF", null)
        }
        if (t.contains("britam")) {
            return Classification("Savings", "Britam EQ and MMF", null)
        }
        if (t.contains("etica")) {
            return Classification("Savings", "Etica MMF", null)
        }
        if (t.contains("ziidi")) {
            return Classification("Savings", "Ziidi MMF", null)
        }
        if (t.contains("cic") || t.contains("mmf") || t.contains("money market")) {
            return Classification("Savings", "Sanlam MMF", null)
        }

        // 3. Matatu Transport Saccos (Expenses -> Fare)
        if (t.contains("super metro") || t.contains("2nk") || t.contains("lopha") || t.contains("metro") || t.contains("kbs") || t.contains("city hoppa") || t.contains("matatu")) {
            return Classification("Expenses", "Fare")
        }

        // 4. Utilities / Bills (Matching verified sheet: Electricity, Water, WIFI, Minutes, Rent, Monthly Shopping)
        if (t.contains("kenya power") || t.contains("kplc") || t.contains("electricity")) {
            return Classification("Bills", "Electricity")
        }
        if (t.contains("nairobi water") || t.contains("water")) {
            return Classification("Bills", "Water")
        }
        if (t.contains("safaricom home") || t.contains("zuku") || t.contains("faiba") || t.contains("wifi") || t.contains("poa") || t.contains("internet")) {
            return Classification("Bills", "WIFI")
        }
        if (t.contains("airtime") || t.contains("minutes")) {
            return Classification("Bills", "Minutes")
        }
        if (t.contains("rent")) {
            return Classification("Bills", "Rent")
        }

        // 5. Bank Transfers / Balance adjustments
        if (t.contains("equity") || t.contains("equity bank")) {
            return Classification("Balance", "", "Equity Bank")
        }
        if (t.contains("i&m") || t.contains("i and m")) {
            return Classification("Balance", "", "I&M Bank")
        }

        // Default P2P or personal expense
        return Classification("Expenses", "House Supplies")
    }

    private fun classifyMerchant(merchant: String): Classification {
        val m = merchant.lowercase(Locale.ROOT)
        // Supermarkets / Shopping -> Bills: Monthly Shopping
        if (m.contains("naivas") || m.contains("carrefour") || m.contains("quickmart") || m.contains("chandarana") || m.contains("cleanshelf")) {
            return Classification("Bills", "Monthly Shopping")
        }
        // Restaurants / Dining -> Expenses: Eating Out
        if (m.contains("kfc") || m.contains("java") || m.contains("artcaffe") || m.contains("pizza") || m.contains("restaurant") || m.contains("cafe") || m.contains("hotel")) {
            return Classification("Expenses", "Eating Out")
        }
        // Matatus / Fare -> Expenses: Fare
        if (m.contains("super metro") || m.contains("2nk") || m.contains("metro") || m.contains("matatu") || m.contains("uber") || m.contains("bolt") || m.contains("fare")) {
            return Classification("Expenses", "Fare")
        }
        // Fuel / Gas -> Expenses: Gas
        if (m.contains("total") || m.contains("shell") || m.contains("rubis") || m.contains("ola") || m.contains("gas") || m.contains("energy")) {
            return Classification("Expenses", "Gas")
        }
        // Groceries / Vegetables -> Expenses: Mama Mboga
        if (m.contains("mboga") || m.contains("grocer") || m.contains("market") || m.contains("veg")) {
            return Classification("Expenses", "Mama Mboga")
        }
        // Barber / Salon -> Expenses: Kinyozi
        if (m.contains("kinyozi") || m.contains("barber") || m.contains("salon")) {
            return Classification("Expenses", "Kinyozi")
        }
        // Clothes / Shoes -> Expenses: Clothes / Shoes
        if (m.contains("shoe") || m.contains("footwear")) {
            return Classification("Expenses", "Shoes")
        }
        if (m.contains("clothe") || m.contains("apparel") || m.contains("boutique")) {
            return Classification("Expenses", "Clothes")
        }
        return Classification("Expenses", "House Supplies")
    }

    private fun formatIsoDate(rawDate: String): String {
        val parts = rawDate.split("/")
        if (parts.size != 3) return SimpleDateFormat("yyyy-MM-dd", Locale.ROOT).format(Date())
        val day = parts[0].padStart(2, '0')
        val month = parts[1].padStart(2, '0')
        var year = parts[2]
        if (year.length == 2) year = "20$year"
        return "$year-$month-$day"
    }
}
